grpclib/protocol.py

```diff
@@ -230,6 +238,65 @@ def close(self) -> None:
             del self._transport
             if hasattr(self._connection, '_frame_dispatch_table'):
                 del self._connection._frame_dispatch_table
+        if self._ping_handle is not None:
+            self._ping_handle.cancel()
+        if self._close_by_ping_handler is not None:
+            self._close_by_ping_handler.cancel()
+
+    async def _ping(self) -> None:
+        if self._config._keepalive_time is None or self._last_received() == 0:
+            self._ping_handle = None
+            return
+
+        if not self._config._keepalive_permit_without_calls:
+            count_of_connection = self.streams_started - self.streams_failed - \
+                                  self.streams_succeeded
+            if count_of_connection == 0:
+                self._ping_handle = None
+                return
+        current_time = time.monotonic()
+        time_to_next_ping = self._config._keepalive_time - (
+                    current_time - self._last_received())
+        while time_to_next_ping > 0:
+            await asyncio.sleep(time_to_next_ping)
+            current_time = time.monotonic()
+            time_to_next_ping = self._config._keepalive_time - (
+                        current_time - self._last_received())
+
+        max_ping_count = self._config._http2_max_pings_without_data
+        while (max_ping_count > 0 or
+               self._config._http2_max_pings_without_data == 0):
+            if self._config._http2_max_pings_without_data != 0:
+                max_ping_count -= 1
+            data = struct.pack('!Q', int(time.monotonic() * 10 ** 6))
+            self._connection.ping(data)
+            self.flush()
+            if self._close_by_ping_handler is None:
+                self._close_by_ping_handler = asyncio.get_event_loop().\
+                    call_later(self._config._keepalive_timeout, self.close)
+            await asyncio.sleep(
+                self._config._http2_min_sent_ping_interval_without_data)
+
+    def data_process(self) -> None:
+        if self._ping_handle is not None:
+            self._ping_handle.cancel()
+        if self._close_by_ping_handler is not None:
+            self._close_by_ping_handler.cancel()
+            self._close_by_ping_handler = None
+        self._ping_handle = asyncio.get_event_loop().create_task(self._ping())
```

Comment:
It seemed to me that frequent calls that do nothing are not very good, and the response time will differ from those set in the parameters up. But considering how much it will be easier. It is a good idea to do so.

Reply:
done